pkg/controller/nodelink/node_link_controller.go

```diff
@@ -0,0 +1,342 @@
+		c.logger.Infof("Error syncing machine %v: %v", key, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	c.logger.Infof("Dropping machine %q out of the queue: %v", key, err)
+	c.queue.Forget(key)
+}
+
+// syncNode will sync the node with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncNode(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing node")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing node")
+	}()
+
+	_, _, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	node, err := c.nodeLister.Get(key)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Info("node has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.processNode(node)
+}
+func (c *Controller) processNode(node *corev1.Node) error {
+	nodeLog := c.logger.WithField("node", node.Name)
+	machineKey, ok := node.Annotations[machineAnnotationKey]
+	// No machine annotation, this is likely the first time we've seen the node,
+	// need to load all machines and search for one with matching IP.
+	var matchingMachine *capiv1.Machine
+	if !ok {
+		// Find this nodes internal IP so we can search for a matching machine:
+		var nodeInternalIP string
+		for _, a := range node.Status.Addresses {
+			if a.Type == corev1.NodeInternalIP {
+				nodeInternalIP = a.Address
+				break
+			}
+		}
+		if nodeInternalIP == "" {
+			nodeLog.Warnf("unable to find InternalIP for node")
+			return fmt.Errorf("unable to find InternalIP for node: %s", node.Name)
+		}
+
+		allMachines, err := c.machinesLister.Machines(kubeClusterNamespace).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		nodeLog.Debugf("node has no machine annotation, searching %d machines for IP match", len(allMachines))
+		for _, m := range allMachines {
+			for _, a := range m.Status.Addresses {
+				// Use the internal IP to look for matches:
+				if a.Type == corev1.NodeInternalIP && a.Address == nodeInternalIP {
+					matchingMachine = m
+					nodeLog.Infof("found matching machine: %s", matchingMachine.Name)
+					break
+				}
+			}
+		}
+	} else {
+		var err error
+		namespace, machineName, err := cache.SplitMetaNamespaceKey(machineKey)
+		if err != nil {
+			nodeLog.Infof("machine annotation format is incorrect %v: %v\n", machineKey, err)
+			return err
+		}
+		matchingMachine, err = c.machinesLister.Machines(namespace).Get(machineName)
+		if err != nil {
+			return err
+		}
+	}
+	if matchingMachine == nil {
+		nodeLog.Warnf("no matching machine found for node")
+		return fmt.Errorf("no matching machine found for node: %s", node.Name)
+	}
+
+	modNode := node.DeepCopy()
+
+	modNode.Annotations["machine"] = fmt.Sprintf("%s/%s", matchingMachine.Namespace, matchingMachine.Name)
+
+	if node.ObjectMeta.Labels == nil {
```

Comment:
really minor nit ... ObjectMeta is not needed because it's embedded anonymously... node.Labels works just as well.